common/djangoapps/student/models.py

```diff
@@ -165,14 +168,39 @@ def anonymous_id_for_user(user, course_id, save=True):
     cached_id = getattr(user, '_anonymous_id', {}).get(course_id)
     if cached_id is not None:
         return cached_id
+    # check if an annonymous id already exists for this user and course_id combination
+    anonymous_user_ids = AnonymousUserId.objects.filter(user=user).filter(course_id=course_id)
```

Comment:
Note from discussion: Separate PR with just this retrieval so we can check for DB impact -- before making any change to algorithm or SECRET_KEY.